File: drivers/nodejs/boyodb/transactions.js
/**
 * Transaction support for BoyoDB Node.js driver.
 *
 * Provides ACID transaction support with savepoints and isolation levels.
 */

/**
 * Isolation levels for transactions.
 * @enum {string}
 */
const IsolationLevel = {
  READ_UNCOMMITTED: 'READ UNCOMMITTED',
  READ_COMMITTED: 'READ COMMITTED',
  REPEATABLE_READ: 'REPEATABLE READ',
  SERIALIZABLE: 'SERIALIZABLE',
  SNAPSHOT: 'SNAPSHOT',
};

/**
 * Transaction wrapper for ACID operations.
 *
 * @example
 * const tx = client.transaction();
 * try {
 *   await tx.begin();
 *   await tx.execute("INSERT INTO users (id, name) VALUES (1, 'Alice')");
 *   await tx.execute("UPDATE accounts SET balance = balance - 100");
 *   await tx.commit();
 * } catch (err) {
 *   await tx.rollback();
 *   throw err;
 * }
 *
 * // Or with auto-commit/rollback:
 * await client.withTransaction(async (tx) => {
 *   await tx.execute("INSERT INTO users (id, name) VALUES (1, 'Alice')");
 *   await tx.execute("UPDATE accounts SET balance = balance - 100");
 * });
 */
class Transaction {
  /**
   * Create a new transaction.
   * @param {Client} client - BoyoDB client
   * @param {Object} [options={}] - Transaction options
   * @param {IsolationLevel} [options.isolationLevel] - Isolation level
   * @param {boolean} [options.readOnly=false] - Read-only transaction
   */
  constructor(client, options = {}) {
    this._client = client;
    this._isolationLevel = options.isolationLevel || null;
    this._readOnly = options.readOnly || false;
    this._active = false;
    this._savepoints = [];
  }

  /**
   * Start the transaction.
   * @returns {Promise<void>}
   */
  async begin() {
    let sql = 'BEGIN';
    if (this._isolationLevel) {
      sql += ` ISOLATION LEVEL ${this._isolationLevel}`;
    }
    if (this._readOnly) {
      sql += ' READ ONLY';
    }

    await this._client.exec(sql);
    this._active = true;
  }

  /**
   * Commit the transaction.
   * @returns {Promise<void>}
   */
  async commit() {
    if (this._active) {
      await this._client.exec('COMMIT');
      this._active = false;
      this._savepoints = [];
    }
  }

  /**
   * Rollback the transaction.
   * @returns {Promise<void>}
   */
  async rollback() {
    if (this._active) {
      await this._client.exec('ROLLBACK');
      this._active = false;
      this._savepoints = [];
    }
  }

  /**
   * Create a savepoint.
   * @param {string} name - Savepoint name
   * @returns {Promise<void>}
   */
  async savepoint(name) {
    if (!this._active) {
      throw new Error('No active transaction');
    }
    await this._client.exec(`SAVEPOINT ${name}`);
    this._savepoints.push(name);
  }

  /**
   * Rollback to a savepoint.
   * @param {string} name - Savepoint name
   * @returns {Promise<void>}
   */
  async rollbackTo(name) {
    if (!this._active) {
      throw new Error('No active transaction');
    }
    await this._client.exec(`ROLLBACK TO SAVEPOINT ${name}`);
    const idx = this._savepoints.indexOf(name);
    if (idx !== -1) {
      this._savepoints = this._savepoints.slice(0, idx + 1);
    }
  }

  /**
   * Release a savepoint.
   * @param {string} name - Savepoint name
   * @returns {Promise<void>}
   */
  async release(name) {
    if (!this._active) {
      throw new Error('No active transaction');
    }
    await this._client.exec(`RELEASE SAVEPOINT ${name}`);
    const idx = this._savepoints.indexOf(name);
    if (idx !== -1) {
      this._savepoints.splice(idx, 1);
    }
  }

  /**
   * Execute SQL within the transaction.
   * @param {string} sql - SQL statement
   * @param {string} [database] - Database to use
   * @returns {Promise<QueryResult>}
   */
  async execute(sql, database) {
    if (!this._active) {
      throw new Error('No active transaction');
    }
    return this._client.query(sql, database);
  }

  /**
   * Check if transaction is active.
   * @returns {boolean}
   */
  get isActive() {
    return this._active;
  }
}

module.exports = {
  IsolationLevel,
  Transaction,
};

File: drivers/python/boyodb/transactions.py
"""
Transaction support for BoyoDB.

Provides ACID transaction support with savepoints and isolation levels.
"""

from enum import Enum
from typing import TYPE_CHECKING, Any, Dict, List, Optional

if TYPE_CHECKING:
    from .client import Client


class IsolationLevel(Enum):
    """Transaction isolation levels."""
    READ_UNCOMMITTED = "READ UNCOMMITTED"
    READ_COMMITTED = "READ COMMITTED"
    REPEATABLE_READ = "REPEATABLE READ"
    SERIALIZABLE = "SERIALIZABLE"
    SNAPSHOT = "SNAPSHOT"


class Transaction:
    """
    Transaction context manager for ACID operations.

    Usage:
        with client.transaction() as tx:
            tx.execute("INSERT INTO users (id, name) VALUES (1, 'Alice')")
            tx.execute("UPDATE accounts SET balance = balance - 100 WHERE id = 1")
            # Automatically commits on success, rolls back on exception

    With savepoints:
        with client.transaction() as tx:
            tx.execute("INSERT INTO orders (id) VALUES (1)")
            tx.savepoint("sp1")
            try:
                tx.execute("INSERT INTO order_items (order_id) VALUES (1)")
            except:
                tx.rollback_to("sp1")
            # Transaction continues
    """

    def __init__(
        self,
        client: "Client",
        isolation_level: Optional[IsolationLevel] = None,
        read_only: bool = False,
    ):
        self._client = client
        self._isolation_level = isolation_level
        self._read_only = read_only
        self._active = False
        self._savepoints: List[str] = []

    def __enter__(self) -> "Transaction":
        self.begin()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        if exc_type is not None:
            self.rollback()
        else:
            self.commit()

    def begin(self) -> None:
        """Start the transaction."""
        sql = "BEGIN"
        if self._isolation_level:
            sql += f" ISOLATION LEVEL {self._isolation_level.value}"
        if self._read_only:
            sql += " READ ONLY"

        self._client.exec(sql)
        self._active = True

    def commit(self) -> None:
        """Commit the transaction."""
        if self._active:
            self._client.exec("COMMIT")
            self._active = False
            self._savepoints.clear()

    def rollback(self) -> None:
        """Rollback the transaction."""
        if self._active:
            self._client.exec("ROLLBACK")
            self._active = False
            self._savepoints.clear()

    def savepoint(self, name: str) -> None:
        """Create a savepoint."""
        if not self._active:
            raise RuntimeError("No active transaction")
        self._client.exec(f"SAVEPOINT {name}")
        self._savepoints.append(name)

    def rollback_to(self, name: str) -> None:
        """Rollback to a savepoint."""
        if not self._active:
            raise RuntimeError("No active transaction")
        self._client.exec(f"ROLLBACK TO SAVEPOINT {name}")
        # Remove savepoints after the one we rolled back to
        if name in self._savepoints:
            idx = self._savepoints.index(name)
            self._savepoints = self._savepoints[:idx + 1]

    def release(self, name: str) -> None:
        """Release a savepoint."""
        if not self._active:
            raise RuntimeError("No active transaction")
        self._client.exec(f"RELEASE SAVEPOINT {name}")
        if name in self._savepoints:
            self._savepoints.remove(name)

    def execute(self, sql: str, database: Optional[str] = None) -> Any:
        """Execute SQL within the transaction."""
        if not self._active:
            raise RuntimeError("No active transaction")
        return self._client.query(sql, database)

    @property
    def is_active(self) -> bool:
        """Check if transaction is active."""
        return self._active
